File: 프로그래머스/lv3/42627. 디스크 컨트롤러/디스크 컨트롤러.js
const solution = (jobs) => {
    let answer = 0,
        j = 0,
        time = 0;
    jobs.sort((a, b) => a[0] - b[0]);

    const priorityQueue = [];
    while (j < jobs.length || priorityQueue.length) {
        if (jobs.length > j && time >= jobs[j][0]) {
            priorityQueue.push(jobs[j++]);
            priorityQueue.sort((a, b) => a[1] - b[1]);
            continue;
        }
        if (priorityQueue.length) {
            time += priorityQueue[0][1];
            answer += time - priorityQueue[0][0];
            priorityQueue.shift();
        } else time = jobs[j][0];
    }
    return parseInt(answer / jobs.length);
}

File: 프로그래머스/3/42627. 디스크 컨트롤러/디스크 컨트롤러.java
import java.util.*;

class Solution {    
    public int solution(int[][] jobs) {
        PriorityQueue<int[]> pg = new PriorityQueue<>((a, b)-> a[1] - b[1]);
        
        Arrays.sort(jobs, (a, b) -> a[0] - b[0]);
        
        int jobIndex = 0;
        int currentTime = 0;
        int totalTime = 0;
        int cnt = 0;
        
        while(cnt < jobs.length) {
            while(jobIndex < jobs.length && currentTime >= jobs[jobIndex][0]) {
                pg.offer(jobs[jobIndex++]);
            }
            
            if(pg.isEmpty()){
                currentTime = jobs[jobIndex][0];
            } else {
                int[] job = pg.poll();
                totalTime += currentTime + job[1] - job[0];
                currentTime += job[1];
                cnt++;
            }
        }
        
        return totalTime / jobs.length;
    }
}